### PR TITLE
Add flash partition capsule support

### PR DESCRIPTION
**Pull request by helloxiling**
This PR introduces the flash partition capsule `flash_partition.rs`, which provides the syscall driver for user space applications to perform read, write, and erase operations on data of arbitrary length from the underlying flash storage.  

The detailed changes include: 
1. Code reorganization: creating a separate package named `flash_driver` to encapsulate the relevant flash driver files. 
2. Definition on the generic interface for reading, writing and erasing the arbitrary length of data on flash storage in `flash_driver/hil.rs`.  
3. Implementation of the flash storage logical driver `flash_storage_to_page.rs` that maps read, write, and erase operations of arbitrary length into page-based operations. This driver is essential for enabling the flash partition capsule. 
4. Addition of in-kernel integration testing `flash_storage_test.rs` for the flash storage logical driver. 
5. Bug fix in `flash_ctrl` driver: ensuring interrupt clearance occurs before the callback, as the callback may initiate another flash operation. 
6. Instantiate one instance of flash partition in board.rs as a placeholder for user mode testing.  
7. Minor update: changing `page_size` from `1024` to `256` in flash_ctrl driver and emulated flash_ctrl to reflect the most common NOR flash configuration.

The user mode testing on flash partition will be added in separate PR.

### CODE REVIEW

File: runtime/flash/src/flash_storage_to_pages.rs [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+
+                    self.page_buffer.replace(page_buffer);
+
+                    self.driver.erase_page(address / page_size)
+                } else {
+                    // This erase is non-page-aligned, so we need to do a read first.
+                    self.address.set(address);
+                    self.remaining_length.set(length);
+
+                    match self.driver.read_page(address / page_size, page_buffer) {
```
Comment (swenson) [inline runtime/flash/src/flash_storage_to_pages.rs:190]: Do we still need to do the erase after this read?

File: runtime/capsules/src/flash_partition.rs [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+        if let Some(processid) = self.current_app.take() {
+            let _ = self.apps.enter(processid, move |_, kernel_data| {
+                // Need to copy in the contents of the buffer
+                let _ = kernel_data
+                    .get_readwrite_processbuffer(rw_allow::READ)
+                    .and_then(|read| {
+                        read.mut_enter(|app_buffer| {
+                            let read_len = cmp::min(app_buffer.len(), length);
+
+                            let d = &app_buffer[0..read_len];
```
[the remark this answers is not in the document]
Comment (helloxiling) [reply runtime/capsules/src/flash_partition.rs:272]: Sure, will update.

File: runtime/capsules/src/flash_partition.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+                                            // Check that the internal buffer and the buffer that was
+                                            // allowed are long enough.
+                                            let write_len =
+                                                cmp::min(active_len, kernel_buffer.len());
+
+                                            let d = &app_buffer[0..write_len];
+                                            for (i, c) in
+                                                kernel_buffer[0..write_len].iter_mut().enumerate()
+                                            {
+                                                *c = d[i].get();
```
[the remark this answers is not in the document]
Comment (helloxiling) [reply runtime/capsules/src/flash_partition.rs:183]: Sure, will refactor it.

File: runtime/flash/src/flash_storage_to_pages.rs (hunk 1 above)
Comment (helloxiling) [reply runtime/flash/src/flash_storage_to_pages.rs:190]: This is the non-page-aligned erase scenario.  We do `read-modify-write` to simulate the erase.  It is handled in the block `state:: Erase` within `read_complete`.  However, if it is an actual flash, we might have to add `erase` in the sequence.

File: runtime/capsules/src/flash_partition.rs [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+use core::cmp;
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::processbuffer::{ReadableProcessBuffer, WriteableProcessBuffer};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::{ErrorCode, ProcessId};
+
+/// Each partition is presented to userspace as a separate driver number.
+/// Below is the temporary driver number for each partition.
+pub const IMAGE_PAR_DRIVER_NUM: usize = 0x5006;
```
[the remark this answers is not in the document]
Comment (helloxiling) [reply runtime/capsules/src/flash_partition.rs:14]: Thanks for catching this.  Sorry it was a typo..  I had intended to assign `0x50006` to flash partition driver just after most storage relevant driver definition.  Is it ok to use `0x5000_0006` ?

File: runtime/flash/src/flash_storage_to_pages.rs [hunk 5]
```diff
@@ -391,6 +386,27 @@ impl<F: hil::flash::Flash> hil::flash::Client<F> for FlashStorageToPages<'_, F>
...
                     .map(move |client| client.erase_done(self.length.get()));
+            } else if self.partial_erase_wb_pending.get() {
+                // Write back the page
+                let page_index = self.address.get() % page_size;
+                // Length is either the rest of the page or how much is left.
+                let len = cmp::min(page_size - page_index, self.remaining_length.get());
+                // Which page was read and which is going to be written back to.
+                let page_number = self.address.get() / page_size;
+
+                self.remaining_length.subtract(len);
```
Comment (swenson) [inline runtime/flash/src/flash_storage_to_pages.rs:397]: Nice. I did not know about `NumericalCellExt`.

**Review by swenson — APPROVED** [review]
LGTM

**Pull request merged**